### PR TITLE
feat(RELEASE-1031): define alerts for Release availability

### PR DESCRIPTION
**Pull request by theflockers**
this commit adds new alerts for a new Release Service SLO namely ReleaseServiceControllerManagerAvailability.

### CODE REVIEW

**Review by yftacherzog — CHANGES_REQUESTED** [review]
Can you add unit tests for the added alert?

**Comment by theflockers** [comment]
> Can you add unit tests for the added alert?

Definitely. I totally forgot about that.

**Comment by theflockers** [comment]
Also this new alert requires the `kube_pod_container_status_ready`, which will be added by the PR https://github.com/redhat-appstudio/infra-deployments/pull/4437/files

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml [hunk 1]
```diff
@@ -59,3 +59,21 @@ spec:
         alert_team_handle: <!subteam^S03SVBS426R>
         runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/release-service/release-processing-latency.md?ref_type=heads
         team: release
+
+    - alert: ReleaseServiceControllerManagerAvailability
+      expr: |
+        rate(
+            sum( (sum(kube_pod_container_status_ready{namespace="release-service", container="manager"}) by(job) OR on() vector(0))) [10m:]
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:66]: What's the reason for having `sum` inside `sum` here?

[the remark this answers is not in the document]
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:66]: the comments a related. This metric give 1 or absent in case of pod not ready. To calculate the rate, I need it to be when absent, so the `vector` needs to be added. But this creates an extra item without labels, for instance:

when the data exists (or with 1)
```
kube_pod_container_status_ready{namespace="release-service"}  1
{}          0
```

About the corner case, that's the one explained  - if no data (which is the case when the pod is down) it gives 0.

I will look up how to emulate this in the test.

[the remark this answers is not in the document]
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:66]: I am trying to simplify the query using other metrics now that they are available.

[the remark this answers is not in the document]
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml]: I fully changed the query to be more intuitive.

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml [hunk 2]
```diff
@@ -59,3 +59,27 @@ spec:
...
         runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/release-service/release-processing-latency.md?ref_type=heads
         team: release
+
+    - alert: ReleaseServiceControllerManagerAvailability
+      expr: |
+        (
+          avg_over_time
+          (
+            (sum(kube_pod_container_status_ready{namespace="release-service", container="manager"}) by(namespace))
+            [1h:]
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:69]: Is it intentional that we have 1h at the numerator and 1d below?

File: test/promql/tests/data_plane/release_service_test.yaml [hunk 3]
```diff
@@ -57,3 +57,24 @@ tests:
               alert_team_handle: <!subteam^S03SVBS426R>
               runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/release-service/release-validation-latency.md?ref_type=heads
               team: release
+
+  - interval: 1m
+    input_series:
+      - series: 'kube_pod_container_status_ready{namespace="release-service", container="manager"}'
```
Comment (yftacherzog) [inline test/promql/tests/data_plane/release_service_test.yaml:63]: We have `sum` on the query, but we only have a single data series on the tests, so we don't test summing.

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml [hunk 4]
```diff
@@ -59,3 +59,27 @@ spec:
...
+          )
+          /
+          max_over_time(sum(kube_pod_container_status_ready{namespace="release-service", container="manager"}) by(namespace)[1d:])
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:72]: If we only keep a single namespace (`namespace="release-service"`), then I think we don't need to sum by namespace, as all data series will have the same namespace anyhow. i.e. I think we can drop the `by(namespace)`. No?

File: test/promql/tests/data_plane/release_service_test.yaml [hunk 5]
```diff
@@ -57,3 +57,24 @@ tests:
               alert_team_handle: <!subteam^S03SVBS426R>
               runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/release-service/release-validation-latency.md?ref_type=heads
               team: release
+
+  - interval: 1m
```
Comment (yftacherzog) [inline test/promql/tests/data_plane/release_service_test.yaml:61]: I suggest adding a test case that will include data series from multiple clusters (i.e. entries with `source_cluster` with varying values). This way the tests can verify that we aggregate what we wanted to aggregate. E.g. you might not want to aggregate time series between different clusters.

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml (hunk 2 above)
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:69]: Yes. The reason is that we might one of the metrics absent for a moment (maybe in the hour it checked it and I the calculation would be done with a missing node. Using one day (theoretically) I can have a  "good enough" assumption of how many nodes I should have up in that hour. I am open to suggestion.

File: test/promql/tests/data_plane/release_service_test.yaml (hunk 5 above)
Comment (theflockers) [reply test/promql/tests/data_plane/release_service_test.yaml:61]: Indeed. It makes  total sense.

File: test/promql/tests/data_plane/release_service_test.yaml (hunk 3 above)
Comment (theflockers) [reply test/promql/tests/data_plane/release_service_test.yaml:63]: indeed.

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml (hunk 4 above)
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:72]: I will check. I tried without it, for something broke. I might have mixed problems when I tried it.

File: test/promql/tests/data_plane/release_service_test.yaml (hunk 5 above)
Comment (theflockers) [reply test/promql/tests/data_plane/release_service_test.yaml:61]: done.

File: test/promql/tests/data_plane/release_service_test.yaml (hunk 3 above)
Comment (theflockers) [reply test/promql/tests/data_plane/release_service_test.yaml]: added samples from different clusters.

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml (hunk 4 above)
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml]: removed the `by(..)`

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml (hunk 2 above)
Comment (yftacherzog) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:69]: I think we should have a test case for that scenario.

File: test/promql/tests/data_plane/release_service_test.yaml (hunk 5 above)
Comment (yftacherzog) [reply test/promql/tests/data_plane/release_service_test.yaml:61]: I think this is currently summing across all clusters. Is that the intention?

Comment (theflockers) [reply test/promql/tests/data_plane/release_service_test.yaml:61]: yes. We consider the service down if the pod is down in any cluster.

File: rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml (hunk 2 above)
Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:69]: added.

Comment (yftacherzog) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:69]: Cool. Can you also add a test case in which the alert is not firing?

Comment (theflockers) [reply rhobs/alerting/data_plane/prometheus.release_service_alerts.yaml:69]: added.

**Review by yftacherzog — APPROVED** [review]
lgtm

**Review by yftacherzog — CHANGES_REQUESTED** [review]
@theflockers Can you squash the commits?

**Comment by theflockers** [comment]
> @theflockers Can you squash the commits?

definitely.

**Review by yftacherzog — APPROVED** [review]
lgtm
